logical/framework/openapi.go

```diff
@@ -0,0 +1,413 @@
+
+			// Set default responses.
+			if len(props.Responses) == 0 {
+				if opType == logical.DeleteOperation {
+					op.Responses["204"] = openapi.StdRespNoContent
+				} else {
+					op.Responses["200"] = openapi.StdRespOK
+				}
+			}
+
+			// Add any defined response details.
+			for code, responses := range props.Responses {
+				var description string
+				content := make(openapi.Content)
+
+				for i, resp := range responses {
+					if i == 0 {
+						description = resp.Description
+					}
+					if resp.Example != nil {
+						mediaType := resp.MediaType
+						if mediaType == "" {
+							mediaType = "application/json"
+						}
+
+						// Only one example per media type is allowed, so first one wins
+						if _, ok := content[mediaType]; !ok {
+							content[mediaType] = &openapi.MediaTypeObject{
+								Schema: &openapi.Schema{
+									Example: cleanResponse(resp.Example),
+								},
+							}
+						}
+					}
+				}
+
+				// a nil pointer when empty is needed for omitempty to work
+				var c *openapi.Content
+				if len(content) > 0 {
+					c = &content
+				}
+				op.Responses[code] = &openapi.Response{
+					Description: description,
+					Content:     c,
+				}
+			}
+
+			switch opType {
+			case logical.CreateOperation, logical.UpdateOperation:
+				pi.Post = op
+			case logical.ReadOperation, logical.ListOperation:
+				pi.Get = op
+			case logical.DeleteOperation:
+				pi.Delete = op
+			}
+		}
+
+		doc.Paths["/"+path] = &pi
+	}
+
+	return nil
+}
+
+func specialPathMatch(path string, specialPaths []string) bool {
+	// Test for exact or prefix match of special paths.
+	for _, sp := range specialPaths {
+		if sp == path ||
+			(strings.HasSuffix(sp, "*") && strings.HasPrefix(path, sp[0:len(sp)-1])) {
+			return true
+		}
+	}
+	return false
+}
+
+// expandPattern expands a regex pattern by generating permutations of any optional parameters
+// and changing named parameters into their {openapi} equivalents.
+func expandPattern(pattern string) []string {
+	var paths []string
+
+	// This construct is added by GenericNameRegex and is much easier to remove now
+	// than to compensate for in the other regexes.
+	pattern = strings.Replace(pattern, `\w(([\w-.]+)?\w)?`, "", -1)
+
+	// Initialize paths with the original pattern or the halves of an
+	// alternation, which is also present in some patterns.
+	matches := altRe.FindAllStringSubmatch(pattern, -1)
+	if len(matches) > 0 {
+		paths = []string{matches[0][1], matches[0][2]}
+	} else {
+		paths = []string{pattern}
+	}
+
+	// Expand all optional regex elements into two paths. This approach is really only useful up to 2 optional
+	// groups, but we probably don't want to deal with the exponential increase beyond that anyway.
+	for i := 0; i < len(paths); i++ {
+		p := paths[i]
+		match := optRe.FindStringIndex(p)
+		if match != nil {
+			paths[i] = p[0:match[0]] + p[match[0]+1:match[1]-2] + p[match[1]:]
+			paths = append(paths, p[0:match[0]]+p[match[1]:])
+			i--
```

Comment:
Clever! Is the length of paths, then, computed only once at the beginning of the loop, rather than each iteration of the loop?

Reply:
It is necessarily computed on every iteration. Each time through we might split an element into two, which is also why there is an index (`i`) and the weird `i--` which is keeping the loop from advancing. I'll probably add some more comments to this method as it's far from obvious what's going on with all those match indices.

Reply:
:+1: